src/operator/tensor/index_update.cu

```diff
@@ -0,0 +1,202 @@
+      }
+      val_id[seg + ind_ndim - 1] = (val_shape[seg + ind_ndim - 1] == 1) ? 0 : i;
+      index_t val_dest = mxnet_op::dot(val_id, val_stride);
+      atomicAdd(&grad_val[val_dest], ograd[id + _i]);
+    }
+  }
+};
+
+template<>
+void IndexUpdateOpBackwardValImpl<gpu>(const OpContext& ctx,
+                                const TBlob& grad_val,
+                                const TBlob& ograd,
+                                const TBlob& t_ind,
+                                const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> ograd_tail_shape,
+                                const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> ograd_pre_stride,
+                                const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> val_stride,
+                                const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> val_shape,
+                                const int tail_size, const int ind_num, const int ind_ndim,
+                                const int ndim) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  mshadow::Stream<gpu> *s = ctx.get_stream<gpu>();
+  int seg = MXNET_SPECIAL_MAX_NDIM - ndim;
+  MSHADOW_TYPE_SWITCH(grad_val.type_flag_, DType, {
+    Kernel<IndexUpdateBackwardValGPUKernel, gpu>::Launch(
+        s, ind_num, grad_val.dptr<DType>(), ograd.dptr<DType>(), t_ind.dptr<int>(),
+        ograd_tail_shape, ograd_pre_stride, val_stride, val_shape, tail_size,
+        ind_num, ind_ndim, ndim, seg);
+  });
+}
+
+template<typename DType>
+struct IndexUpdateBackwardAGPUKernel {
+  MSHADOW_XINLINE static void Map(size_t i, DType* out_grad,
+                                  const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> grada_pre_stride,
+                                  const int tail_size, const int ind_num, const int ind_ndim,
+                                  const int32_t* ind, const int seg,
+                                  const int req) {
+    index_t id = 0;
+    for (int dim = 0; dim < ind_ndim; ++dim) {
+      id += grada_pre_stride[seg + dim] * ind[dim * ind_num + i];
+    }
+    id *= tail_size;
+    for (int _i = 0; _i < tail_size; ++_i) {
+      out_grad[id + _i] = 0;
+    }
+  }
+};
+
+template<>
+void IndexUpdateOpBackwardAImpl<gpu>(const OpContext& ctx,
+                                     const TBlob& grad_a,
+                                     const TBlob& ograd,
+                                     const TBlob& ind,
+                                     const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> grada_pre_stride,
+                                     const int tail_size, const int ind_num, const int ind_ndim,
+                                     const int seg, const int req) {
+  using namespace mxnet_op;
+  using namespace mshadow;
+  mshadow::Stream<gpu> *s = ctx.get_stream<gpu>();
+  MSHADOW_TYPE_SWITCH(grad_a.type_flag_, DType, {
+    size_t temp_mem_size = ind.shape_.Size() * sizeof(int) +
+                           ograd.shape_.Size() * sizeof(DType);
```

Comment:
Here, we mix `int` type and `DType`. The second pointer may not be appropriately aligned with `DType`. You may consider to revise the code and add the `PadBytes`: https://github.com/apache/incubator-mxnet/blob/1b02225fefd8ccc93bc73223f0d3cde103fad661/src/operator/tensor/ordering_op-inl.h#L350-L351

The reason to do this is that CUDA forces memory alignment, i.e., we will always have the following assertion:

```c++
DType* ptr;
ASSERT static_cast<size_t>(ptr) % sizeof(DType) == 0.
```

By default, calling cudaMalloc will give you a pointer with at least 256-bytes aligned address. However, after shifting the pointer to `ptr += ind.shape_.Size() * sizeof(int)`, the new address may not be aligned with `sizeof(DType)`.

Reply:
OK, I have revised this segment in [https://github.com/apache/incubator-mxnet/pull/18545/commits/c251837d2703813a254fd75cc426d50139220272](url)

Reply:
OK, I have revised the code in commit [https://github.com/apache/incubator-mxnet/pull/18545/commits/c251837d2703813a254fd75cc426d50139220272](url)